test/integration/int_server.h

```diff
@@ -0,0 +1,451 @@
+ private:
+  ServerFilterChain(const ServerFilterChain &) = delete;
+  ServerFilterChain &operator=(const ServerFilterChain &) = delete;
+
+  Envoy::Network::TransportSocketFactory &transport_socket_factory_;
+  std::vector<Envoy::Network::FilterFactoryCb> network_filter_factories_;
+};
+
+/**
+ * A convenience class for creating a listening socket bound to localhost
+ */
+class LocalListenSocket : public Envoy::Network::TcpListenSocket {
+ public:
+  /**
+   * Create a listening socket bound to localhost.
+   *
+   * @param ip_version v4 or v6.  v4 by default.
+   * @param port the port.  If 0, let the kernel allocate an avaiable ephemeral
+   * port.  0 by default.
+   * @param options socket options.  nullptr by default
+   * @param bind_to_port if true immediately bind to the port, allocating one if
+   * necessary.  true by default.
+   */
+  LocalListenSocket(
+      Envoy::Network::Address::IpVersion ip_version =
+          Envoy::Network::Address::IpVersion::v4,
+      uint16_t port = 0,
+      const Envoy::Network::Socket::OptionsSharedPtr &options = nullptr,
+      bool bind_to_port = true);
+
+  virtual ~LocalListenSocket();
+
+  LocalListenSocket(LocalListenSocket &&) = default;
+  LocalListenSocket &operator=(LocalListenSocket &&) = default;
+
+ private:
+  LocalListenSocket(const LocalListenSocket &) = delete;
+  void operator=(const LocalListenSocket &) = delete;
+};
+
+/**
+ * A convenience class for passing callbacks to a Server.  If no callbacks are
+ * provided, default callbacks that track some simple metrics will be used.   If
+ * callbacks are provided, they will be wrapped with callbacks that maintain the
+ * same simple set of metrics.
+ */
+class ServerCallbackHelper {
+ public:
+  ServerCallbackHelper(ServerRequestCallback request_callback = nullptr,
+                       ServerAcceptCallback accept_callback = nullptr,
+                       ServerCloseCallback close_callback = nullptr);
+
+  virtual ~ServerCallbackHelper();
+
+  ServerCallbackHelper(ServerCallbackHelper &&) = default;
+  ServerCallbackHelper &operator=(ServerCallbackHelper &&) = default;
+
+  uint32_t connectionsAccepted() const;
+  uint32_t requestsReceived() const;
+  uint32_t localCloses() const;
+  uint32_t remoteCloses() const;
+  ServerAcceptCallback acceptCallback() const;
+  ServerRequestCallback requestCallback() const;
+  ServerCloseCallback closeCallback() const;
+
+  /*
+   * Wait until the server has accepted n connections and seen them closed (due
+   * to error or client close)
+   */
+  void wait(uint32_t connections);
+
+  /*
+   * Wait until the server has seen a close for every connection it has
+   * accepted.
+   */
+  void wait();
+
+ private:
+  ServerCallbackHelper(const ServerCallbackHelper &) = delete;
+  void operator=(const ServerCallbackHelper &) = delete;
+
+  ServerAcceptCallback accept_callback_;
+  ServerRequestCallback request_callback_;
+  ServerCloseCallback close_callback_;
+
+  std::atomic<uint32_t> accepts_{0};
```

Comment:
I think the google idiom of using Mutex and Mutex::Await along with lock annotations is simpler and less error prone.  Then you don't need to explicitly have a std::condition_variable or use atomics.

Reply:
Resolved because the discussion moved to another conversation close to the tricky part of the code